edit/applies-to-line-widget.js

```diff
@@ -488,40 +488,77 @@ function createAppliesToLineWidget(cm) {
     };
   }
 
-  function *findAppliesTo(posStart, posEnd) {
-    const text = cm.getValue();
-    const re = /^[\t ]*@-moz-document[\s\n]+/gm;
-    const applyRe = new RegExp([
-      /(?:\/\*[\s\S]*?(?:\*\/\s*|$))*/,
-      /(url|url-prefix|domain|regexp)/,
-      /\(((['"])(?:\\\\|\\\n|\\\3|[^\n])*?\3|[^)\n]*)\)\s*(,\s*)?/,
-    ].map(rx => rx.source).join(''), 'giy');
-    let match;
-    re.lastIndex = posStart;
-    while ((match = re.exec(text))) {
-      if (match.index >= posEnd) {
-        return;
-      }
+  function *findAppliesTo(posStart, posEnd, lineIndexes) {
+    const funcRe = /^(url|url-prefix|domain|regexp)$/i;
+    let pos;
+    const eatToken = sticky => {
+      if (!sticky) skipSpace(pos, posEnd);
+      pos.ch++;
+      const token = cm.getTokenAt(pos);
+      pos.ch = token.end;
+      return CodeMirror.cmpPos(pos, posEnd) <= 0 ? token : {};
+    };
+    const docCur = cm.getSearchCursor('@-moz-document', posStart);
+    while (docCur.findNext() &&
+           CodeMirror.cmpPos(docCur.pos.to, posEnd) <= 0) {
+      if (!/\bdef\b/.test(cm.getTokenTypeAt(docCur.pos.from))) continue;
       const applies = [];
-      let m;
-      applyRe.lastIndex = re.lastIndex;
-      while ((m = applyRe.exec(text))) {
+      pos = docCur.pos.to;
+      do {
+        skipSpace(pos, posEnd);
+        const funcIndex = lineIndexes[pos.line] + pos.ch;
+        const func = eatToken().string;
+        // no space allowed before the opening parenthesis
+        if (!funcRe.test(func) || eatToken(true).string !== '(') break;
+        const url = eatToken();
+        if (url.type !== 'string' || eatToken().string !== ')') break;
+        const unquotedUrl = unquote(url.string);
         const apply = createApply(
-          m.index,
-          m[1],
-          unquote(m[2]),
-          unquote(m[2]) !== m[2]
+          funcIndex,
+          func,
+          unquotedUrl,
+          unquotedUrl !== url.string
         );
         applies.push(apply);
-        re.lastIndex = applyRe.lastIndex;
-      }
+      } while (eatToken().string === ',');
       yield {
-        pos: cm.posFromIndex(match.index),
+        pos: docCur.pos.from,
         applies
       };
     }
   }
 
+  function skipSpace(pos, posEnd) {
+    let {ch, line} = pos;
+    let lookForEnd;
+    line--;
+    cm.doc.iter(pos.line, posEnd.line + 1, ({text}) => {
+      line++;
+      while (true) {
+        if (lookForEnd) {
```

Comment:
It is not unpredictable. At the worst case, the above pattern `(?:\s+|\/\*.*?\*\/|(\/\*.*))+` backtrack the entire line once so it is as fast as linear search.

---

The original pattern causes catastrophic backtracking because of this pattern:
```js
/(\/\*[\s\S]*?(\*\/\s*|$))*(url|url-prefix|domain|regexp)/
```
When the second group is not found, the comment part will be expanded. It can be fixed by changing the first part to something like:
```js
/(?:\/\*(?:[^*/]|\*(?!\/)|\/)*\*\/|\s)*/
```

Reply:
It is unpredictable because lines can be extremely long in some styles.

Reply:
The thing is, unless you can re-write the entire original regexp (and keep the original code) and guarantee superquick execution, there's no point in rewriting skipSpace() function - it's already quite obvious and simple, whereas the regexps you propose aren't really readable and won't really simplify the code.

Reply:
> It is unpredictable because lines can be extremely long in some styles.

When I said the performance is predictable, I meant that the relationship between the time and the data size (the length of the text) is linear, hence it is the same as the current implementation.

> simplify the code.

I tended to avoid the `while(true)` loop, multiple `indexOf` call in a single line, and `includes` when the regex can handle them.

If you hate regular expression and do want to keep the current logic. At least you should change the pattern to:
```js
/(?:\s+|\/\*)+/y
```
Because `\s*` always matches an empty string.

Reply:
Yeah, removed it.

Reply:
Personally I love regexps and for my personal stuff I occasionally write some super-optimized and scary monsters, but I definitely don't think it's a good idea to use them in a public project that's not dedicated to showcasing our regexp skills. As for indexOf or includes, these are cheap performance-wise, anyway, so further work would be micro-optimization, I think.